README.md
# MotorCortex Counter

## Installation
```bash
$ npm install @kissmybutton/motorcortex-counter
# OR
$ yarn add @kissmybutton/motorcortex-counter
```

```javascript
import Counter from "@kissmybutton/motorcortex-counter";
```


## Key Concepts / Features


## Browser compatibility 
| Chrome | Safari | IE / Edge | Firefox | Opera |
| --- | --- | --- | --- | --- |
| 24+ | 6+ | 10+ | 32+ | 15+ |

## Documentation

### Import and load the plugin to MotorCortex
```javascript
import MotorCortex from "@kissmybutton/motorcortex";
import Counter from "@kissmybutton/motorcortex-counter";

const Counter = MotorCortex.loadPlugin(AnimePluginDefinition);
```

### Create a Counter Incident to Animate a Number

```javascript
import MotorCortex from "@kissmybutton/motorcortex";
import AnimePluginDefinition from "@kissmybutton/motorcortex-anime";

const AnimePlugin = MotorCortex.loadPlugin(AnimePluginDefinition);

const MyClip = new MotorCortex.Clip({
   html: `
        <div id="number-counter">
            <p class="effect-target">100</p>
            <p>%</p>
        </div>
    `,
  css: `
   #number-counter {
        width: 60%;
        height: 60%;
        display: flex;
        justify-content: center;
        align-items: center;
        font-size: 800%;    
    }
  `
});

const counterEffect = new Counter.Counter({
    animatedAttrs: {
        count: 100,
    },
    initialValues: {
        count: 0,
    },
    decimals: 2,
}, {
    selector: '.effect-target',
    duration: 1000,
    easing: "easeInOutQuad",
});

MyClip.addIncident(counterEffect, 1000);

MyClip.play();
```

With this Incident you are able to animate and customize the progress of a number from an initial value to the target value. The "count" parameter allows you to set that initial and target value, and as long as the input is a "Number" there are no other limitations. The "decimals" parameter is optional. It defines the decimal accurary of the counter. Its input must be an integer number between 0 and 20. 



### Exposed Incidents
* `Counter`

### Demo
https://kissmybutton.github.io/motorcortex-counter/demo/


## License
[MIT License](https://opensource.org/licenses/MIT)


  
  
[![Kiss My Button](https://presskit.kissmybutton.gr/logos/kissmybutton-logo-small.png)](https://kissmybutton.gr)


CHANGELOG.md
# Changelog

### [1.0.1](https://www.github.com/kissmybutton/motorcortex-counter/compare/v1.0.0...v1.0.1) (2021-02-09)


### Bug Fixes

* version and release bump ([e5ad677](https://www.github.com/kissmybutton/motorcortex-counter/commit/e5ad677b6e3e7f2ddc083f0d947b7cdd2413a514))

## 1.0.0 (2021-02-09)


### Bug Fixes

* version & package.json update ([38793f1](https://www.github.com/kissmybutton/motorcortex-counter/commit/38793f1940ca6b107903ed7aeeaaecaed5729919))


src/Incidents/Counter.js
import MotorCortex from '@kissmybutton/motorcortex';

/**
 * The purpose of Effects is to timely alter the state or value of attributes of
 * selected elements of the context, specified on the "selector"
 * property of theirs.
 *
 * The attributes of the elements that the Effect manipulates are 
 * always defined on the attrs.animatedAttrs object, passed to it on its constructor. 
 * Each key of this object corresponds to an attribute that the Effect will alter and the value
 * of each specifies the final value to go to.
 * On runtime, each Effect is analysed first by element and secondly 
 * by animatedAttr. 
 * For example an Effect that has the selector ".my-class",
 * that applies in two elements of the context, and has two animatedAttrs 
 * will be analysed into four in total "MonoIncidents" (2 elements * 2 animatedAttrs).
 * Each of these produced MonoIncidents refer to a very specific element and
 * to a very specific animated attribute.
 * The Class that you are defining here extends Effect which represents exactly this MonoIncident.
 * 
 * Thus, here you'll find:
 * the following properties:
 * - this.element: provides a reference to the specific element of the MonoIncident
 * - this.attributeKey: the key of the animatedAttr of the MonoIncident
 * - this.targetValue: the final value of the animatedAttr
 * - this.initialValue: the initial value of the animatedAttr
 * and the following methods:
 * - onGetContext
 * - getScratchValue
 * - onProgress
 * which are analysed more inline
 *
 **/
export default class Counter extends MotorCortex.Effect{
    /**
    * the very first MonoIncident of the specific element and the 
    * specific attribute that will ever enter a Clip will be asked
    * to provide the initial (the scratch) value of its animatedAttr
    * for its element.  
    **/
    getScratchValue(){
        return 0;
    }
    
    /**
    * The moment the Effect gets applied as MonoIncident to the specific
    * element and for the specific animatedAttr.
    * You can use this method to initialise anything you need to initialise
    * in order to use it on the onProgress method
    **/
    onGetContext(){
        this.element.innerHTML = this.initialValue;
    }

    /**
    * Takes two arguments the "fraction" which is a number from 0 to 1, representing
    * the fraction (the percentage) of the duration that we are in,
    * and the millisecond which defines the absolute millisecond.
    * You can use this method to animate your attribute.
    * Remember that you don't need to worry about easings. Easings are already 
    * applied before reaching the execution of this method. This method's
    * arguments have already been re-calculated based on the easing.
    **/
    onProgress(fraction, millisecond){
        let currentVal = this.initialValue + ((this.targetValue - this.initialValue) * fraction);

        if (this.attrs.decimals) {
            currentVal = currentVal.toFixed(this.attrs.decimals);
        } else {
            currentVal = Math.trunc(currentVal);
        }

        this.element.innerHTML = currentVal;
    }
}


demo/index.js
import MotorCortex from '@kissmybutton/motorcortex';
import MCCounter from "../dist/bundle.umd";
const Counter = MotorCortex.loadPlugin(MCCounter);

import Player from "@kissmybutton/motorcortex-player";


const clip = new MotorCortex.HTMLClip({
    html: `
        <div class="container">
            <div id="number-counter">
                <p class="effect-target">100</p>
                <p>%</p>
            </div>
        </div>
    `,
    css: `
        .container {
            width: 100%;
            height: 100%;
            display: flex;
            justify-content: center;
            align-items: center;
        }
        #number-counter {
            width: 60%;
            height: 60%;
            display: flex;
            justify-content: center;
            align-items: center;
            font-size: 800%;
        }
    `,
    host: document.getElementById('clip'),
    containerParams: {
        width: '100%',
        height: '100%'
    }
});

const counterEffect = new Counter.Counter({
    animatedAttrs: {
        count: 110,
    },
    initialValues: {
        count: 16,
    },
}, {
    selector: '.effect-target',
    duration: 1000,
    easing: "easeInQuad",
});
clip.addIncident(counterEffect, 0);


const counterEffect2 = new Counter.Counter({
    animatedAttrs: {
        count: 50,
    },
    decimals: 2,
}, {
    selector: '.effect-target',
    duration: 1000,
    easing: "easeInOutQuad",
});

clip.addIncident(counterEffect2, 1000);

const player = new Player({clip, timeFormat: 'ms'});